undate/dateformat/base.py

```diff
@@ -0,0 +1,20 @@
+# base class for date format parsers
+
+
+class BaseDateFormat:
+    """Base class for parsing and formatting dates for specific formats."""
+
+    # Subclasses should define a unique name.
+    name = "Base Formatter"
+
+    def parse(self, value):
+        # should return an undate or undate interval
+        raise NotImplementedError
+
+    def to_string(self, undate):
+        # convert an undate or interval to string representation for this format
+        raise NotImplementedError
+
+    @classmethod
+    def available_formatters(cls):
+        return {c.name: c for c in cls.__subclasses__()}
```

Comment:
Tests seem to be fine, but I'm encountering an error with this when trying Undate in the Python CLI. For `cls.__subclassess__()` to work, the subclass definitions have to be available (subclass module has to be imported). This works:
```
>>> from undate.dateformat.iso8601 import ISO8601DateFormat
>>> from undate.undate import Undate, UndateInterval 
>>> str(Undate(2022, 4))
'2022-04'
>>> from undate.dateformat.base import BaseDateFormat
>>> base = BaseDateFormat()
>>> base.available_formatters()
{'ISO8601': <class 'undate.dateformat.iso8601.ISO8601DateFormat'>}
```
But this does not:
```
>>> from undate.dateformat.base import BaseDateFormat
>>> base = BaseDateFormat()
>>> base.available_formatters()
{}
>>> from undate.undate import Undate, UndateInterval
>>> str(Undate(2022, 4))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/colecrawford/GitHub/hackathon-2022/undate/undate.py", line 42, in __init__
    formatter = BaseDateFormat.available_formatters()[self.DEFAULT_FORMAT]()
KeyError: 'ISO8601'
```

Reply:
oh, good catch! I've run into that before, it's easy to forget / harder to trigger in unit tests. (We should figure out how to duplicate this problem in the tests, though)

What if we import everything into the main init and document that you should import undate from there? Or I could look at how I've solved it before...

Reply:
If you have a good pattern for that we could borrow, that would be great. If we import everything we may run into circular import issues?

Reply:
Yeah, I was proposing using the init to avoid circular imports. I took a look, the example I was thinking of was django-specific and doesn't really help us (I have a note in the comments to import in app ready method to make sure they are loaded if you want to use the signal handler)

However, here's what I propose, which I think will work: 
-  in `undate.dateformat.__init__.py` we should import the base date format and 8601 subclass
-  in `undate.undate` we import base date format from the module (`from undate.dateformat import BaseDateFormat`)
- document how to add a new formatter and include the instructions that it must be added to the imports in `undate.dateformat.__init__.py`